tests/unit/unit-util.cpp

```diff
@@ -229,6 +229,53 @@ TEST_P(UnitUtilDimacsTests, cnf_to_dimacs)
 }
 
 
+
+TEST_P(UnitUtilDimacsTests, tseitin)
+{
+  s->set_opt("incremental", "true");
+  Term p = s->make_symbol("p", boolsort);
+  Term q = s->make_symbol("q", boolsort);
+  Term r = s->make_symbol("r", boolsort);
+  Term t = s->make_symbol("t", boolsort);
+  
+  //a=((p or q) and r) implies (not t)
+  Term a = s->make_term(Implies, s->make_term(And, s->make_term(Or, p, q), r), s->make_term(Not, t));
+  Term cnf1 = to_cnf(a, s);
+  string st1 = cnf1->to_string();
+  string ans1 =
+      "(and cnf_formula_new_4 (and (or (not t) (not cnf_formula_new_1)) (or t "
+      "cnf_formula_new_1) (or (not cnf_formula_new_2) p q) (and (or "
+      "cnf_formula_new_2 (not p)) (or cnf_formula_new_2 (not q))) (or "
+      "cnf_formula_new_3 (not cnf_formula_new_2) (not r)) (and (or "
+      "cnf_formula_new_2 (not cnf_formula_new_3)) (or r (not "
+      "cnf_formula_new_3))) (or (or (not cnf_formula_new_3) cnf_formula_new_1) "
+      "(not cnf_formula_new_4)) (or cnf_formula_new_3 cnf_formula_new_4) (or "
+      "(not cnf_formula_new_1) cnf_formula_new_4)))";
+  ASSERT_TRUE(ans1 == st1)<<s->get_solver_enum();
+  //b=(not (p xor q))
+  Term b = s->make_term(Not, s->make_term(Xor, p, q));
+  Term cnf2 = to_cnf(b, s);
+  string st2 = cnf2->to_string();
+  string ans2 =
+      "(and cnf_formula_new_6 (and (or (or (not p) (not q)) (not "
+      "cnf_formula_new_5)) (or (or p q) (not cnf_formula_new_5)) (or (or "
+      "cnf_formula_new_5 q) (not p)) (or (or cnf_formula_new_5 p) (not q)) (or "
+      "(not cnf_formula_new_5) (not cnf_formula_new_6)) (or cnf_formula_new_5 "
+      "cnf_formula_new_6)))";
+  ASSERT_TRUE(ans2 == st2);
+  //c=((not p) and p)
+  Term c = s->make_term(And, s->make_term(Not, p), p);
+  Term cnf3 = to_cnf(c, s);
+  string st3 = cnf3->to_string();
+  string ans3 =
+      "(and cnf_formula_new_8 (and (or (not p) (not cnf_formula_new_7)) (or p "
+      "cnf_formula_new_7) (or cnf_formula_new_8 (not cnf_formula_new_7) (not "
+      "p)) (and (or cnf_formula_new_7 (not cnf_formula_new_8)) (or p (not "
+      "cnf_formula_new_8)))))";
+  ASSERT_TRUE(ans3 == st3)<<ans3<<endl<<endl<<s->get_solver_enum();
+}
+
+
```

Comment:
Let's add some more simple tests and corner cases as inputs:
1. An input that is already in cnf
2. the constants `true/false`
3. A literal
4. A clause
5. A formula that contains something that is not a Boolean variable (then make sure that an exception is throen).

Reply:
Ok sure.
1. The formula which I would output would be different, is that fine? I few extra variables will be added.
2. Do you mean a large formula contains within itself true/false. As of now, I have not handled this case but I can implement it.

Reply:
Well, as an optimization, I suggest you add a check before transforming, and if the input is already in CNF, you can just return the input itself. 

I suggest handling true/false, they do occur in practical encodings, so supporting it is desirable. With CNF, they can be simulated, e.g., by adding a unit clause with a single fresh variable and another unit clause with a single negated fresh variable, and these can represent the constants.

And for the test, I recommend testing both the simple formulas `true` and `false` as well as their incorporation in other formulas.

Reply:
Ok, could you also check out my reply to the comment regarding asserting strings or terms?

Reply:
Yes, I did!